File: python/src/workshop_api/external_invoice_provider/invoice_provider_store.py
from workshop_api.external_invoice_provider.invoice_provider_schemas import (
    InvoiceProviderResponse,
    InvoiceProviderUpsertRequest,
)


class InvoiceProviderStore:
    def __init__(self) -> None:
        self._invoices: dict[str, InvoiceProviderResponse] = {}

    def find_all(self) -> list[InvoiceProviderResponse]:
        return list(self._invoices.values())

    def find_by_id(self, invoice_id: str) -> InvoiceProviderResponse | None:
        return self._invoices.get(invoice_id)

    def save(
        self,
        invoice_id: str,
        request: InvoiceProviderUpsertRequest,
    ) -> InvoiceProviderResponse:
        response = InvoiceProviderResponse(
            invoiceId=invoice_id,
            customerReference=request.customer_reference,
            contractReference=request.contract_reference,
            amountInCents=request.amount_in_cents,
            currency=request.currency,
            dueDate=request.due_date,
            status=request.status,
            description=request.description,
            externalCorrelationId=request.external_correlation_id,
            metadata=request.metadata,
        )
        self._invoices[invoice_id] = response
        return response

    def save_response(
        self,
        response: InvoiceProviderResponse,
    ) -> InvoiceProviderResponse:
        self._invoices[response.invoice_id] = response
        return response

    def delete(self, invoice_id: str) -> None:
        self._invoices.pop(invoice_id, None)

    def clear(self) -> None:
        self._invoices.clear()

File: java/src/main/java/com/workshop/architecture/external_invoice_provider/InvoiceProviderStore.java
package com.workshop.architecture.external_invoice_provider;

import org.springframework.stereotype.Component;

import java.util.*;

@Component
public class InvoiceProviderStore {

    private final Map<String, InvoiceProviderResponse> invoices = new LinkedHashMap<>();

    public List<InvoiceProviderResponse> findAll() {
        return new ArrayList<>(invoices.values());
    }

    public Optional<InvoiceProviderResponse> findById(String invoiceId) {
        return Optional.ofNullable(invoices.get(invoiceId));
    }

    public InvoiceProviderResponse save(String invoiceId, InvoiceProviderUpsertRequest request) {
        InvoiceProviderResponse response = new InvoiceProviderResponse(
                invoiceId,
                request.customerReference(),
                request.contractReference(),
                request.amountInCents(),
                request.currency(),
                request.dueDate(),
                request.status(),
                request.description(),
                request.externalCorrelationId(),
                request.metadata()
        );
        invoices.put(invoiceId, response);
        return response;
    }

    public InvoiceProviderResponse save(InvoiceProviderResponse response) {
        invoices.put(response.invoiceId(), response);
        return response;
    }

    public void delete(String invoiceId) {
        invoices.remove(invoiceId);
    }

    public void clear() {
        invoices.clear();
    }
}
